### DISCUSSION TITLE
[SDK] Transaction hash mismatch from Provider.sendTransaction

### DISCUSSION THREAD

**Discussion by everdimension** [zkSync-Community-Hub/zksync-developers#508] [ZK Stack] [answered]
### Environment

Testnet

### Acknowledgement

- [X] I understand that issues are expected with the Go SDK as they are being updated.

### Issue Description

I have submitted a transaction using the Provider that's exported from `zksync-ethers`
The node request worked ok, but the library threw this message when it got the response:

_Error: Transaction hash mismatch from Provider.sendTransaction. 
(expectedHash="0xab2fe90cb484c37cde6ebc890a8906e55459d2786b2a8d3da02cca77c71769c0", returnedHash="0x4d23f8115cfc16cb40e18f7d4105ef8029ddbe1508171f02c950bb44c90b7d8e", code=UNKNOWN_ERROR, version=providers/5.7.2)_

Does that mean that the RPC url returned something wrong?
The hash seems to work: https://explorer.zksync.caldera.dev/tx/0x4d23f8115cfc16cb40e18f7d4105ef8029ddbe1508171f02c950bb44c90b7d8e

---
My package versions:
```
zksync-ethers v5.7.0
node v20.10.0
npm v10.2.5
```

### Expected Behavior

I guess `Provider.sendTransaction` resolves with the returned hash

### Code Example

My code looks roughly like this:

```
import { EIP712Signer, utils as zkSyncUtils, Provider as ZksProvider } from 'zksync-ethers';

async sendTransaction(transaction) {
  const provider =  new ZksProvider(nodeUrl);
  const { chainId } = transaction;
  invariant(chainId, 'ChainId missing from TransactionRequest');
  const typedData = createTypedData({
    chainId: normalizeChainId(chainId),
    message: EIP712Signer.getSignInput(transaction),
  });
  const signature = await this.signTypedData_v4({
    context,
    params: { typedData },
  });
  const rawTransaction = zkSyncUtils.serialize(transaction, signature);
  return await this.sendSignedTransaction(rawTransaction);
}

async sendSignedTransaction(serialized: string) {
  const provider =  new ZksProvider(nodeUrl);
  try {
    return await provider.sendTransaction(serialized);
  } catch (error) {
    throw getEthersError(error);
  }
}
```

**Comment by bxpana** [comment]
Thanks for the post, will check things out and get back to you.

**Reply by everdimension** [reply]
@bxpana Any update here? :)

**Reply by dutterbutter** [reply]
Hey 👋 

Can you provide a link to the repo to reproduce?

**Reply by everdimension** [reply]
@dutterbutter Not really. Isn't the code example and the block explorer link enough here?

**Reply by dutterbutter** [reply]
Correct me if I am wrong but it seems you are using your own node? The tx hash is not present on Era mainnet or Sepolia. Given this is not reproducible there I was curious to see if its specific to the node in which is being interacted with (e.g. nodeURL).

**Reply by everdimension** [reply]
Yes, I'm using this node: https://rpc.zksync.caldera.dev/
I asked Caldera developers and they don't know about this error, and since this check is done on the side of zksync library, I assume it's appropriate to ask here.

Is it the node doing something wrong? What does the hash check look like on your side?

**Reply by everdimension** [reply]
@dutterbutter ping :) Can you please take a look?

**Comment by everdimension** [comment, marked as answer]
I think I found what needed to be changed.

You need to attach the signature differently.
Here's approximate diff:

```diff
  const signature = await this.signTypedData_v4({
    context,
    params: { typedData },
  });
+ transaction.customData.customSignature = signature;
- const rawTransaction = zkSyncUtils.serialize(transaction, signature);
+ const rawTransaction = zkSyncUtils.serialize(transaction);
  return this.sendSignedTransaction(rawTransaction);
```

Here's a reference in the zksync docs: https://docs.zksync.io/build/developer-reference/account-abstraction.html#sending-transactions-from-an-account

**Reply by bxpana** [reply]
Great to hear! So things are working for you now?

**Reply by everdimension** [reply]
Yes, seems so.
